### PR TITLE
Delete discovery gateway

### PR DESCRIPTION
**Pull request by systay**
## Description

This pull request removes the `DiscoveryGateway` struct and related files. The `DiscoveryGateway` has been our legacy gateway since the implementation of the new `tabletgateway` through https://github.com/vitessio/vitess/pull/6155.

## Related Issue(s)
- Contributes to #9218 

## Checklist
- [ ] Should this PR be backported? _NO_
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/executor_framework_test.go [hunk 1]
```diff
@@ -496,6 +457,8 @@ func createExecutorEnv() (executor *Executor, sbc1, sbc2, sbclookup *sandboxconn
 	_ = hc.AddTestTablet(cell, "a0-c0", 1, "TestExecutor", "a0-c0", topodatapb.TabletType_PRIMARY, true, 1, nil)
 	_ = hc.AddTestTablet(cell, "c0-e0", 1, "TestExecutor", "c0-e0", topodatapb.TabletType_PRIMARY, true, 1, nil)
 	_ = hc.AddTestTablet(cell, "e0-", 1, "TestExecutor", "e0-", topodatapb.TabletType_PRIMARY, true, 1, nil)
+	// Below is needed so that SendAnyWherePlan doesn't fail
+	_ = hc.AddTestTablet(cell, "e0-", 1, "TestXBadVSchema", "-20", topodatapb.TabletType_PRIMARY, true, 1, nil)
```
Comment (harshit-gangal) [inline go/vt/vtgate/executor_framework_test.go:461]: the problem seems to be related to hostname, if you change the hostname from `e0-` to `random` the test starts working.
The bug may be a test thing when the hostname is the same.

Comment (deepthi) [reply go/vt/vtgate/executor_framework_test.go:461]: ah good point. we should change `e0-` to `-20` to match the shard name.

Comment (deepthi) [reply go/vt/vtgate/executor_framework_test.go:461]: `-20` also conflicts. I guess we should make it `0`.
I tried that, the next error is 
```
--- FAIL: TestExecutorMaxPayloadSizeExceeded (0.00s)
    executor_test.go:1992:
        	Error Trace:	executor_test.go:1992
        	Error:      	Not equal:
        	            	expected: <nil>(<nil>)
        	            	actual  : *vterrors.fundamental(Code: NOT_FOUND
        	            	tablet aa-0000000010 not found
        	            	)
        	Test:       	TestExecutorMaxPayloadSizeExceeded
        	Messages:   	err should be nil
```

File: go/vt/vtgate/gateway.go [hunk 2]
```diff
@@ -35,10 +34,10 @@ import (
...
-	bufferImplementation  = flag.String("buffer_implementation", "keyspace_events", "Allowed values: healthcheck (legacy implementation), keyspace_events (default)")
-	initialTabletTimeout  = flag.Duration("gateway_initial_tablet_timeout", 30*time.Second, "At startup, the gateway will wait up to that duration to get one tablet per keyspace/shard/tablettype")
+	// Deprecated GatewayImplementation allows you to choose which gateway to use for vtgate routing. Defaults to tabletgateway, other option is discoverygateway
+	_                    = flag.String("gateway_implementation", "tabletgateway", "Deprecated. The only available gateway_implementation is tabletgateway")
+	bufferImplementation = flag.String("buffer_implementation", "keyspace_events", "Allowed values: healthcheck (legacy implementation), keyspace_events (default)")
+	initialTabletTimeout = flag.Duration("gateway_initial_tablet_timeout", 30*time.Second, "At startup, the gateway will wait up to that duration to get one tablet per keyspace/shard/tablettype")
 	// RetryCount is the number of times a query will be retried on error
 	// Make this unexported after DiscoveryGateway is deprecated
```
Comment (deepthi) [inline go/vt/vtgate/gateway.go:42]: We can change `RetryCount` -> `retryCount` and remove this comment.

File: go/vt/vtgate/executor_dml_test.go [hunk 3]
```diff
@@ -409,7 +409,7 @@ func TestDeleteEqual(t *testing.T) {
 }
 
 func TestUpdateScatter(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (deepthi) [reply go/vt/vtgate/executor_dml_test.go:411]: What I did instead is to not initialize `primarySession` when it is declared, but create it in each of the test env functions - `createExecutorEnv` etc. That seems to fix all the failures in this particular test.

**Comment by frouioui** [comment]
I will open a follow-up pull request rebased on top of this one to remove the `Gateway` interface introduced by https://github.com/vitessio/vitess/pull/5992.

File: go/vt/discovery/fake_healthcheck.go [hunk 4]
```diff
@@ -55,9 +51,9 @@ func NewFakeHealthCheck(ch chan *TabletHealth) *FakeHealthCheck {
 // FakeHealthCheck implements discovery.HealthCheck.
 type FakeHealthCheck struct {
 	// mu protects the items map
-	mu    sync.RWMutex
-	items map[string]*fhcItem
-
+	mu               sync.RWMutex
+	items            map[string]*fhcItem
+	currentTabletUID sync2.AtomicInt32
```
Comment (systay) [inline go/vt/discovery/fake_healthcheck.go:56]: nit: couldn't we use a normal int and use the mutex to protect it?

Comment (frouioui) [reply go/vt/discovery/fake_healthcheck.go:56]: it is not a bad idea, wherever we use `currentTabletUID` (in `Reset` & `AddFakeTablet`) we acquire the mutex's lock, I'll change that to an `int`

Comment (frouioui) [reply go/vt/discovery/fake_healthcheck.go]: fixed via 4b16187ea7

**Review by systay — COMMENTED** [review]
I opened this PR, so I can't accept it, but it LGTM ⭐ 

Removing old code is the best ❤️

**Comment by frouioui** [comment]
@deepthi in `healthcheck.go`  you left a ToDo to deprecate some variables. I am not sure which ones you were talking about. Can you guide me?

Link to the corresponding ToDo:
https://cs.github.com/vitessio/vitess/blob/605beab9015ce69840f014cabd98a8ecae4d9b31/go/vt/discovery/healthcheck.go#L70

**Comment by deepthi** [comment]
> @deepthi in `healthcheck.go` you left a ToDo to deprecate some variables. I am not sure which ones you were talking about. Can you guide me?
> 
> Link to the corresponding ToDo: https://cs.github.com/vitessio/vitess/blob/605beab9015ce69840f014cabd98a8ecae4d9b31/go/vt/discovery/healthcheck.go#L70

Not deprecate, just change from exported -> not. All 6 flags below that comment should be reviewed to see if they are still being used outside `discovery` package. If not, we can change them back.

**Review by frouioui — APPROVED** [review]

**Pull request merged**